src/zalsa.rs

```diff
@@ -176,9 +176,14 @@ impl Zalsa {
     }
 
     /// Returns the [`MemoTable`][] for the salsa struct with the given id
-    pub(crate) fn memo_table_for(&self, id: Id) -> &MemoTable {
+    pub(crate) fn memo_table_for(&self, id: Id) -> MemoTableWithTypes<'_> {
+        let table = self.table();
         // SAFETY: We are supply the correct current revision
-        unsafe { self.table().memos(id, self.current_revision()) }
+        let (memos, ingredient) =
+            unsafe { table.memos_and_ingredient(id, self.current_revision()) };
+        let types = self.lookup_ingredient(ingredient).memo_table_types();
+        // SAFETY: The memos and the types are from the same ingredient.
+        unsafe { types.attach_memos(memos) }
```

Comment:
I imagine this is where the perf regression comes from, `memo_table_for` is a very hot path and it now does a fair bit more work.